### PR TITLE
Accept correction files produced by ppx_driver

### PR DESCRIPTION
**Pull request by diml**
This PR makes `[@@deriving_inline ...]` works with jbuilder. It requires: https://github.com/janestreet/ppx_driver/pull/17

### CODE REVIEW

File: src/super_context.ml [hunk 1]
```diff
@@ -901,6 +902,24 @@ module PP = struct
...
+    | Some _ | None -> false
+
+  let promote_correction ~uses_ppx_driver fn build =
+    if not uses_ppx_driver then
+      build
+    else
+      Build.progn
+        [ build
+        ; Build.return
+            (A.promote_if
```
Comment (rgrinberg) [inline src/super_context.ml]: Is it really worth introducing a module alias just for this one use?

File: src/super_context.ml [hunk 2]
```diff
@@ -901,6 +902,24 @@ module PP = struct
           mli) in
     { m with impl ; intf }
 
+  let uses_ppx_driver ~pps =
+    match Option.map ~f:Pp.to_string (List.last pps) with
+    | Some "ppx_driver.runner" -> true
+    | Some _ | None -> false
+
+  let promote_correction ~uses_ppx_driver fn build =
+    if not uses_ppx_driver then
```
Comment (rgrinberg) [inline src/super_context.ml]: Small nit: I find else clauses in `if (not ...)` confusing and prefer `if not x then foo else bar` to turn to `if x then bar else foo`.

File: src/action.ml [hunk 3]
```diff
@@ -744,7 +745,8 @@ let rec exec t ~ectx ~dir ~env_extra ~stdout_to ~stderr_to =
         match mode with
         | Always -> files
         | If_corrected_file_exists ->
-          List.filter files ~f:(fun file -> Path.exists file.Promote.src)
+          List.filter files ~f:(fun { Promote. src; dst } ->
+            Path.exists src && Path.exists dst)
```
Comment (rgrinberg) [inline src/action.ml]: Why do we need to make sure that the destination exists now?

Comment (diml) [reply src/action.ml]: Basically we must not promote the file if it's generated. But maybe we should do that check in super_context.ml instead

File: src/super_context.ml (hunk 1 above)
Comment (diml) [reply src/super_context.ml]: It's just that super_context defines an Action module as well, so we get a name clash

File: src/super_context.ml (hunk 2 above)
Comment (diml) [reply src/super_context.ml]: Personally, I find if/match expressions easier to read when the simple choice is first. That's why I wrote it this way

File: src/action.ml (hunk 3 above)
Comment (diml) [reply src/action.ml]: Thinking about this more, doing nothing when the source file doesn't exist seems wrong. For instance if the user generates a file containing some `[@@deriving_inline ...]` attribute, then we at least want to check that they are correct.

I guess `promote-if` could have the following behavior: print the diff but only promote if the source file exist as well.

**Comment by diml** [comment]
I did the change for promote-if.

I also adapted the code so that `-diff-cmd -` is passed only if ppx_driver supports it. To do that ppx_driver installs a file `config.sexp` that contains `has-diff-cmd-dash`.

**Comment by diml** [comment]
actually that's not necessary, the worse that can happen is that ppx_driver will try to run `- file1 file2` and we'll get a weird weeor

**Comment by diml** [comment]
We can merge this now then, I submitted a new minor release of ppx_driver: https://github.com/ocaml/opam-repository/pull/11236

**Pull request merged**